### ISSUE TITLE
Support for ppc/ppc64/s390x/riscv32/riscv64/mips , Closure errors when building V

### ISSUE THREAD

**Issue by xet7** [vlang/v#18737]
### Describe the bug

Building V at Ubuntu 23.04 s390x Docker container, running on RHEL 8.8 s390x. Same error when building V at Ubuntu and RHEL. 

Related to what I'm trying to build at s390x https://github.com/wekan/wekan/wiki/s390x

```
[linux1@wekanrhel84 v]$ uname -a
Linux wekanrhel84 4.18.0-477.13.1.el8_8.s390x #1 SMP Thu May 18 09:10:47 EDT 2023 s390x s390x s390x GNU/Linux
[linux1@wekanrhel84 v]$ cat /etc/redhat-release
Red Hat Enterprise Linux release 8.8 (Ootpa)
[linux1@wekanrhel84 v]$ sudo su
[root@wekanrhel84 v]$ cd
[root@wekanrhel84 ~]# docker run --name ubuntu -h 23.04 -e LANG=C.UTF-8 -it ubuntu:23.04 /bin/bash -l
root@23:/home/ubuntu/repos/v# uname -a
Linux 23.04 4.18.0-477.13.1.el8_8.s390x #1 SMP Thu May 18 09:10:47 EDT 2023 s390x s390x s390x GNU/Linux

root@23:/home/ubuntu/repos/v# make
cd ./vc && git clean -xf && git pull --quiet
cd ./thirdparty/tcc && git clean -xf && git pull --quiet
The executable './thirdparty/tcc/tcc.exe' does not work.
cc  -std=gnu99 -w -o v1.exe ./vc/v.c -lm -lpthread 
./vc/v.c: In function ‘__closure_alloc’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_alloc’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c:16637:51: note: each undeclared identifier is reported only once for each function it appears in
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_init’:
./vc/v.c:16785:30: error: ‘__CLOSURE_GET_DATA_BYTES’ undeclared (first use in this function); did you mean ‘__CLOSURE_GET_DATA’?
16785 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
      |                              __CLOSURE_GET_DATA
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16788:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16788 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_create’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16800:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16800 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
make: *** [GNUmakefile:113: all] Error 1
```
When I try to build TCC from source:
```
root@23:/home/ubuntu/repos/tinycc# ./configure --prefix=/var/tmp/tcc --crtprefix=/var/tmp/tcc/lib:/usr/lib64:/usr/lib/s390x-linux-gnu --libpaths=/var/tmp/tcc/lib:/usr/lib/s390x-linux-gnu:/usr/lib64:/usr/lib:/lib/s390x-linux-gnu:/lib:/usr/local/lib/s390x-linux-gnu:/usr/local/lib  --debug
Unsupported CPU
```

### Expected Behavior

Building V works.

### Current Behavior

Building V did not work, shows closure errors.

### Reproduction Steps

SSH to Linux s390x server.
```
git clone https://github.com/vlang/v

cd v

make
```

### V version

Newest commit after v0.4, commit 049d6853b70be5adf69fc37d98883efe4c695152 changelog: add a missing link

### Environment details (OS name and version, etc.)

Running at IBM LinuxOne Community Cloud, s390x https://github.com/wekan/wekan/wiki/s390x
```
[linux1@wekanrhel84 v]$ uname -a
Linux wekanrhel84 4.18.0-477.13.1.el8_8.s390x #1 SMP Thu May 18 09:10:47 EDT 2023 s390x s390x s390x GNU/Linux
[linux1@wekanrhel84 v]$ cat /etc/redhat-release
Red Hat Enterprise Linux release 8.8 (Ootpa)
[linux1@wekanrhel84 v]$ sudo su
[root@wekanrhel84 v]$ cd
[root@wekanrhel84 ~]# docker run --name ubuntu -h 23.04 -e LANG=C.UTF-8 -it ubuntu:23.04 /bin/bash -l
root@23:/home/ubuntu/repos/v# uname -a
Linux 23.04 4.18.0-477.13.1.el8_8.s390x #1 SMP Thu May 18 09:10:47 EDT 2023 s390x s390x s390x GNU/Linux
```

**Comment by JalonSolov** [comment]
Which C compiler do you have, and what version is it?

**Comment by xet7** [comment]
@JalonSolov 

Ubuntu 23.04 s390x:
```
root@23:/# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/s390x-linux-gnu/12/lto-wrapper
Target: s390x-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 12.2.0-17ubuntu1' --with-bugurl=file:///usr/share/doc/gcc-12/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-12 --program-prefix=s390x-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch=z13 --with-tune=z16 --enable-s390-excess-float-precision --with-long-double-128 --enable-multilib --enable-checking=release --build=s390x-linux-gnu --host=s390x-linux-gnu --target=s390x-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.2.0 (Ubuntu 12.2.0-17ubuntu1)
```

RHEL s390x:
```
[linux1@wekanrhel84 ~]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/s390x-redhat-linux/8/lto-wrapper
Target: s390x-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --disable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --disable-libmpx --enable-gnu-indirect-function --with-long-double-128 --with-arch=z13 --with-tune=z14 --enable-decimal-float --build=s390x-redhat-linux
Thread model: posix
gcc version 8.5.0 20210514 (Red Hat 8.5.0-18) (GCC)
```

**Comment by JalonSolov** [comment]
The Ubuntu version of gcc is fine.  The RHEL version is too old... think you need at least v9 of gcc.

Did you clone the latest version of V?  It is up to v0.4 as of today.

**Comment by xet7** [comment]
@JalonSolov 

Yes. With newest commit on Ubuntu, I get this:
```
root@23:/home/ubuntu/repos/v# make
cd ./vc && git clean -xf && git pull --quiet
cd ./thirdparty/tcc && git clean -xf && git pull --quiet
The executable './thirdparty/tcc/tcc.exe' does not work.
cc  -std=gnu99 -w -o v1.exe ./vc/v.c -lm -lpthread 
./vc/v.c: In function ‘__closure_alloc’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_alloc’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c:16637:51: note: each undeclared identifier is reported only once for each function it appears in
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_init’:
./vc/v.c:16785:30: error: ‘__CLOSURE_GET_DATA_BYTES’ undeclared (first use in this function); did you mean ‘__CLOSURE_GET_DATA’?
16785 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
      |                              __CLOSURE_GET_DATA
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16788:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16788 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_create’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16800:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16800 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
make: *** [GNUmakefile:113: all] Error 1
```

**Comment by xet7** [comment]
After make clean I get this:
```
root@23:/home/ubuntu/repos/v# make clean
rm -rf ./thirdparty/tcc
rm -rf ./thirdparty/legacy
rm -rf ./vc
root@23:/home/ubuntu/repos/v# make
make fresh_vc
make[1]: Entering directory '/home/ubuntu/repos/v'
rm -rf ./vc
git clone --filter=blob:none --quiet https://github.com/vlang/vc ./vc
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 1), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), 926.11 KiB | 16.25 MiB/s, done.
Resolving deltas: 100% (1/1), done.
make[1]: Leaving directory '/home/ubuntu/repos/v'
cd ./vc && git clean -xf && git pull --quiet
make fresh_tcc
make[1]: Entering directory '/home/ubuntu/repos/v'
rm -rf ./thirdparty/tcc
Pre-built TCC not available for thirdparty-linux-s390x at https://github.com/vlang/tccbin, will use the system compiler: cc
git clone --filter=blob:none --quiet --branch thirdparty-unknown-unknown https://github.com/vlang/tccbin ./thirdparty/tcc
remote: Enumerating objects: 1, done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 1
Receiving objects: 100% (1/1), 149 bytes | 149.00 KiB/s, done.
make[2]: Entering directory '/home/ubuntu/repos/v'
The executable './thirdparty/tcc/tcc.exe' does not work.
make[2]: Leaving directory '/home/ubuntu/repos/v'
make[1]: Leaving directory '/home/ubuntu/repos/v'
cd ./thirdparty/tcc && git clean -xf && git pull --quiet
The executable './thirdparty/tcc/tcc.exe' does not work.
cc  -std=gnu99 -w -o v1.exe ./vc/v.c -lm -lpthread 
./vc/v.c: In function ‘__closure_alloc’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_alloc’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c:16637:51: note: each undeclared identifier is reported only once for each function it appears in
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16747:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16747 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_init’:
./vc/v.c:16785:30: error: ‘__CLOSURE_GET_DATA_BYTES’ undeclared (first use in this function); did you mean ‘__CLOSURE_GET_DATA’?
16785 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
      |                              __CLOSURE_GET_DATA
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16788:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16788 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_create’:
./vc/v.c:16637:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16637 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16800:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16800 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
make: *** [GNUmakefile:113: all] Error 1
``

**Comment by xet7** [comment]
This is newest commit:
```
commit 049d6853b70be5adf69fc37d98883efe4c695152 (HEAD -> master, origin/master, origin/HEAD)
Author: Alexander Medvednikov <alexander@vlang.io>
Date:   Sat Jul 1 14:07:51 2023 +0300

    changelog: add a missing link
``

**Comment by JalonSolov** [comment]
I just tried the same thing:  cloned V into an ubuntu:23.04 image, and ran `make`, and it worked as expected.  However, I have a much newer kernel on my system.
```
root@23:/tmp/v# make
make fresh_vc
make[1]: Entering directory '/tmp/v'
rm -rf ./vc
git clone --filter=blob:none --quiet https://github.com/vlang/vc ./vc
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 1), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), 926.11 KiB | 7.35 MiB/s, done.
Resolving deltas: 100% (1/1), done.
make[1]: Leaving directory '/tmp/v'
cd ./vc && git clean -xf && git pull --quiet
make fresh_tcc
make[1]: Entering directory '/tmp/v'
rm -rf ./thirdparty/tcc
git clone --filter=blob:none --quiet --branch thirdparty-linux-amd64 https://github.com/vlang/tccbin ./thirdparty/tcc
remote: Enumerating objects: 25, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 25 (delta 0), reused 3 (delta 0), pack-reused 7
Receiving objects: 100% (25/25), 2.07 MiB | 5.25 MiB/s, done.
make[2]: Entering directory '/tmp/v'
make[2]: Leaving directory '/tmp/v'
make[1]: Leaving directory '/tmp/v'
cd ./thirdparty/tcc && git clean -xf && git pull --quiet
cc  -std=gnu99 -w -o v1.exe ./vc/v.c -lm -lpthread 
./v1.exe -no-parallel -o v2.exe  cmd/v
./v2.exe -nocache -o ./v  cmd/v
rm -rf v1.exe v2.exe
Your `tcc` is working. Good - it is much faster at compiling C source code.
V has been successfully built
V 0.4.0 049d685
root@23:/tmp/v# uname -a
Linux 23.04 6.3.5-2-MANJARO #1 SMP PREEMPT_DYNAMIC Sun Jun  4 18:26:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
root@23:/tmp/v# 
```

**Comment by xet7** [comment]
Hmm, I'll try to create some full VM that could use newer kernel. Building v does work for me on x86_64, but not on s390x with that older kernel.

**Comment by xet7** [comment]
@JalonSolov 

When I run virt-manager KVM VM with Alpine Linux s390x, that has Linux kernel 6.1, I still get closure errors:

![v-alpine-s390x](https://github.com/vlang/v/assets/15545/779b72ae-ff35-4af9-b2f0-ec299b2eefe0)

**Comment by JalonSolov** [comment]
I'm at a loss, then.  Someone else will have to chime in.

**Comment by xet7** [comment]
Hmm, maybe that closure is not defined for Linux s390x CPU, at architecture specific code?

https://github.com/vlang/v/blob/master/vlib/v/gen/c/cheaders.v

**Comment by xet7** [comment]
Also thinking, how to find page size for s390x ? 
```
static int _V_page_size = 0x4000; // 16K
```
https://github.com/vlang/v/blob/master/vlib/v/gen/c/cheaders.v#L79

**Comment by xet7** [comment]
```
$ getconf PAGESIZE
4096
```

**Comment by JalonSolov** [comment]
Ah, you're right... closures aren't defined for all CPUs, as they have to have a tiny bit of assembly.  We would have to add a chunk for s390 for this to work.

**Comment by xet7** [comment]
@JalonSolov 

How can I find what kind of assembly is needed? Does it require compiling C and looking what assembly is created? Or something else?

**Comment by JalonSolov** [comment]
It was hand-written.  Look in `vlib/v/gen/c/cheaders.v` starting on line 64, at the function `c_closure_helpers`.  Looks like closures are currently supported on x86, x86_64, arm32, arm64, RISC-V 32, and RISC-V 64.

@spaceface777 and/or @spytheman might be able to help further.

**Comment by xet7** [comment]
I do not know is this in any way related to this closure support:

https://gcc.gnu.org/pipermail/java-patches/2002q3/005159.html

**Comment by barracuda156** [comment]
Same error on macOS ppc: https://github.com/vlang/v/issues/20507

**Comment by barracuda156** [comment]
> It was hand-written. Look in `vlib/v/gen/c/cheaders.v` starting on line 64, at the function `c_closure_helpers`. Looks like closures are currently supported on x86, x86_64, arm32, arm64, RISC-V 32, and RISC-V 64.
> 
> @spaceface777 and/or @spytheman might be able to help further.

Would be cool if someone adds support here for archs which are missing: PowerPC (ppc / ppc64), s390x, mips.

**Comment by barracuda156** [comment]
@xet7 Somewhat earlier versions somehow worked, despite no support for closures for the arch in question. I have 2023.15 built on macOS PowerPC, while the current version fails exactly like for you.

**Comment by JalonSolov** [comment]
The only way to get this fixed on these extra architectures is to give someone else to have access to systems with the same CPUs, or for you to provide login access on yours.

Then, of course, whoever it is will have to understand those systems well enough to add the appropriate code.

**Comment by barracuda156** [comment]
@JalonSolov Well, I won’t mind giving access to one of my machines, I do not know how to do it, given that there is no static IP and I do not control access point. But there should be some way, of course.

The first thing could be to fix what was broken after 0.3.4, since that version builds perfectly fine with no hacks, despite a lack of arch-specific code.

**Comment by JalonSolov** [comment]
How can that even be tested without having access to the machine where it breaks?  Either you'll have to create the fix, or someone who knows PowerPC well enough will have to do so.

**Comment by xet7** [comment]
@JalonSolov 

One way is to test with Qemu. Here is example how I crosscompiled and tested at many OS/CPU some C89 code. I have tested that those binaries also work at those bare metal architectures that I have available.

https://github.com/xet7/darkesthour

**Comment by xet7** [comment]
@JalonSolov 

Problem is that closure CPU-specific assembler code https://github.com/vlang/v/issues/18737#issuecomment-1616186719

Just a moment, I'll test newest code.

**Comment by JalonSolov** [comment]
Yes, qemu would be one way to get access to at least some of the arch's mentioned, but then which versions of which OSes?

**Comment by xet7** [comment]
@JalonSolov 

Qemu-user runs Linux binaries for many CPUs, no OS install needed:

https://github.com/xet7/darkesthour/blob/main/Source/menu.sh#L436

Deps are here:

https://github.com/xet7/darkesthour/blob/main/Source/deps.sh#L10

I would think, that that try some of the existing assembler code, what parts also works at other CPUs

**Comment by barracuda156** [comment]
> @JalonSolov 
> 
> 
> 
> Problem is that closure CPU-specific assembler code https://github.com/vlang/v/issues/18737#issuecomment-1616186719
> 
> 
> 
> Just a moment, I'll test newest code.

@xet7 While that definitely is the problem, vlang 0.3.4 still builds fine. Something got broken to the effect of making the build impossible. (Presumably it was never optimal, but it worked.)

**Comment by xet7** [comment]
@JalonSolov 

Alternatively, would it be possible to rewrite CPU-specific assembler code for example to C89, so that it would compile for all CPUs.

**Comment by barracuda156** [comment]
> Yes, qemu would be one way to get access to at least some of the arch's mentioned, but then which versions of which OSes?  And that still doesn't address needing someone familiar enough with the assembly code for each of those arch's to be able to handle appropriate changes.

I am somewhat familiar with PowerPC assembler and Darwin ABI, not on a level of writing implementation from scratch, but if we get a Linux version (which is trivial to test), I can adapt it to Darwin and test there.

**Comment by xet7** [comment]
At s390x, with newest commit at master, I get this:

```
~/repos/v$ make
cd ./vc && git clean -xf && git pull --quiet
remote: Enumerating objects: 2, done.
remote: Counting objects: 100% (2/2), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 2 (delta 1), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (2/2), 1017.28 KiB | 13.38 MiB/s, done.
Resolving deltas: 100% (1/1), done.
cd ./thirdparty/tcc && git clean -xf && git pull --quiet
The executable './thirdparty/tcc/tcc.exe' does not work.
cc  -std=gnu99 -w -o v1.exe ./vc/v.c -lm -lpthread 
./vc/v.c: In function ‘__closure_alloc’:
./vc/v.c:16610:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_alloc’?
16610 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16720:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16720 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c:16610:51: note: each undeclared identifier is reported only once for each function it appears in
16610 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16720:40: note: in expansion of macro ‘_CLOSURE_SIZE’
16720 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_init’:
./vc/v.c:16758:30: error: ‘__CLOSURE_GET_DATA_BYTES’ undeclared (first use in this function); did you mean ‘__CLOSURE_GET_DATA’?
16758 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
      |                              __CLOSURE_GET_DATA
./vc/v.c:16610:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16610 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16761:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16761 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
./vc/v.c: In function ‘__closure_create’:
./vc/v.c:16610:51: error: ‘__closure_thunk’ undeclared (first use in this function); did you mean ‘__closure_init’?
16610 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
./vc/v.c:16773:25: note: in expansion of macro ‘_CLOSURE_SIZE’
16773 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
make: *** [GNUmakefile:113: all] Error 1
```

**Comment by JalonSolov** [comment]
... and that error makes perfect sense.  `__closure_thunk` is the assembly piece, and it won't be defined for your arch.  The reason it started failing after 0.3.4 is likely that V got more strict on checking the arch type, and didn't generate a thunk at all unless it matched an arch it knew, where before that it may have generated one by default.

I'd be willing to bet that if you actually tried _using_ a closure with the older version, it would crash the program, since it won't have the correct code.

**Comment by barracuda156** [comment]
@JalonSolov Well, then it needs to be written, whether in assembler or at least a generic fallback in C.

**Comment by barracuda156** [comment]
@felipensp Do we have this somewhere in documentation in assembler commands with registers form rather than encoded? I am not familiar with x86 assembler.

**Comment by JalonSolov** [comment]
The x86 code is shown as comments after the hex values.  Reducing to just the comments shows
```
static const char __closure_thunk[] = {
	// movq  xmm15, QWORD PTR [rip - userdata]
	// jmp  QWORD PTR [rip - fn]
};
static char __CLOSURE_GET_DATA_BYTES[] = {
	// movq rax, xmm15
	// ret
};
```

**Comment by barracuda156** [comment]
@JalonSolov @felipensp Could you please also let us know what will be a simple test-case to determine it works as intended, provided the build itself gets fixed?

(I am not programming in `v`, so if there is an existing procedure to evaluate whether the discussed code works, it gonna be very helpful.)

**Comment by barracuda156** [comment]
@spaceface777 Just in case tagging, since you have added RISC-V implementation of closures.

**Comment by JalonSolov** [comment]
The simplest test would probably be to make your changes, then run `v test vlib/v/tests/closure_test.v` and see if everything passes.  There are other closure tests in `vlib/v/tests` as well, with names starting with `closure_`.

**Comment by barracuda156** [comment]
@xet7 By the way, maybe change the title? The issue is not related to the OS, but rather to certain archs lacking support.

**Comment by tankf33der** [comment]
@xet7 - `s390x` cpu support is done, please try again and report.

**Comment by xet7** [comment]
@tankf33der 

Thanks a lot ! This works great at:

1) s390x Ubuntu 24.10 using Linux kernel 6.11.0-21-generic

2) s390x RHEL 9.5 (Plow) using Linux kernel 5.14.0-503.34.1.el9_5.s390x

**Comment by barracuda156** [comment]
@xet7 We do not have PowerPC and MIPS still, so perhaps this should be re-opened. (Otherwise we could make a new issue, of course.)

**Comment by JalonSolov** [comment]
It is usually better to have separate issues for separate systems, especially since different people may work on them.

**Comment by xet7** [comment]
I'll try how this works at those other CPUs.

**Comment by tankf33der** [comment]
> I'll try how this works at those other CPUs.

riscv32 and 64 already in the code base, others not.

p.s. add sparc64 to list of not implemented platforms.

**Comment by tankf33der** [comment]
> [@tankf33der](https://github.com/tankf33der)
> 
> Thanks a lot ! This works great at:
> 
> 1. s390x Ubuntu 24.10 using Linux kernel 6.11.0-21-generic
> 2. s390x RHEL 9.5 (Plow) using Linux kernel 5.14.0-503.34.1.el9_5.s390x
> 
> Compiling v with make works, compiling examples like vweb/veb/sqlite works, etc. Thanks!

did you try on real hardware or qemu?

**Comment by xet7** [comment]
I'm getting this error at riscv64 SiPeed LicheePi4A that is running Debian trixie/sid with kernel 5.10113-lpi4a 2024.02.29.14.50+3eee2bf4d SMP PREEMPT:

```
/repos/v/examples/veb$ v -g run verb_example.v
failed thidparty object build vmd:
'/home/sipeed/repos/v/thirdparty/tcc/tcc.exe' -std=gnu99
-D_DEFAULT_SOURCE -fwrapv -g -fPIC -I
"/home/sipeed/repos/v/thirdparty/cJSON" -o
"/home/sipeed/.vmodules/.cache/35/
357daef8c03b3f0e2f73a7d0d171cd94.module.json.o" -c 
"/home/sipeed/repos/v/thirdparty/cJSON/cJSON.c"
```
Also, with this I get illegal instruction:
```
v hello_world.v
```

**Comment by tankf33der** [comment]
`riscv64` works successfully in docker+qemu on Alpine.

**Comment by barracuda156** [comment]
> `riscv64` works successfully in docker+qemu on Alpine.

Given a pathetic state of Qemu on powerpc, I won’t be surprised that it can’t emulate real-life riscv cpus, especially given that they vary substantially.

**Comment by tankf33der** [comment]
> I'm getting this error at riscv64 SiPeed LicheePi4A that is running Debian trixie/sid with kernel 5.10113-lpi4a 2024.02.29.14.50+3eee2bf4d SMP PREEMPT:
> 
> ```
> /repos/v/examples/veb$ v -g run veb_example.v
> failed thidparty object build vmd:
> '/home/sipeed/repos/v/thirdparty/tcc/tcc.exe' -std=gnu99
> -D_DEFAULT_SOURCE -fwrapv -g -fPIC -I
> "/home/sipeed/repos/v/thirdparty/cJSON" -o
> "/home/sipeed/.vmodules/.cache/35/
> 357daef8c03b3f0e2f73a7d0d171cd94.module.json.o" -c 
> "/home/sipeed/repos/v/thirdparty/cJSON/cJSON.c"
> ```
> 
> Also, with this I get illegal instruction:
> 
> ```
> v hello_world.v
> ```

@xet7 

run this on latest vlang (if possible) and reply:
```
$ cd v
$ gcc -std=gnu11 -w -o v1 vc/v.c -lm -lpthread
$ ./v1 -version -v
```

it `v1` returns version then closure works correctly (in general) and problem is somewhere else

**Comment by xet7** [comment]
@tankf33der 

## 1) RISCV64 LicheePi4a

riscv64 SiPeed LicheePi4A that is running Debian trixie/sid with kernel 5.10113-lpi4a 2024.02.29.14.50+3eee2bf4d SMP PREEMPT:

```
./v1 version -v

V 0.4.10 87f92b1

cd examples

../v bug hello_world.v

cannot compile /home/sipeed/repos/v/cmd/tools/vbug.v: 1

cc: /usr/bin/ld: target /home/sipeed/repos/v/thirdparty/tcc/lib/libgc.a
not found: File or directory does not exist

cc: collect2: error: ld returned 1 exit status
```

## 2) PPC PowerBook G4

At Mac PowerBook G4, running https://t2sde.org Linux kernel 6.12.5-t2 SMP Wed Dev 18 21:35:20 CET ppc GNU/Linux

Downloaded from https://dl.t2sde.org/binary/2024/?C=M;O=D

https://dl.t2sde.org/binary/2024/t2-24.12-ppc-base-wayland-glibc-gcc-603.iso

https://dl.t2sde.org/binary/2024/t2-24.12-ppc-base-wayland-glibc-gcc-603.sha256

Partitioning: https://t2sde.org/handbook/html/t2.partition.apple.html

![Image](https://github.com/user-attachments/assets/9ba64ae5-f435-465a-85fe-a53810be6dd9)

**Comment by tankf33der** [comment]
> [@tankf33der](https://github.com/tankf33der)
> 
> ## 1) RISCV64 LicheePi4a
> riscv64 SiPeed LicheePi4A that is running Debian trixie/sid with kernel 5.10113-lpi4a 2024.02.29.14.50+3eee2bf4d SMP PREEMPT:
> 
> ```
> ./v1 version -v
> 
> V 0.4.10 87f92b1
> 
> cd examples
> 
> ../v bug hello_world.v
> 
> cannot compile /home/sipeed/repos/v/cmd/tools/vbug.v: 1
> 
> cc: /usr/bin/ld: target /home/sipeed/repos/v/thirdparty/tcc/lib/libgc.a
> not found: File or directory does not exist
> 
> cc: collect2: error: ld returned 1 exit status
> ```
> 

looks riscv64 works on your hardware, this is good.

Rest are debian issue(s). Lets try this:

```
$ apt install libgc-dev
$ export VFLAGS="-cc gcc -d dynamic_boehm"
$ v run hello_world.v
```

**Comment by xet7** [comment]
@tankf33der 

Thanks! Now hello world and veb example works!

**Comment by xet7** [comment]
@tankf33der 

Yes, I can run any program at s390x etc.

I'll try to look is sparc64 available .

**Comment by xet7** [comment]
@tankf33der 

```
linux1@ubu2410s390x:~/vtest$ v version
V 0.4.10 3b4c016
linux1@ubu2410s390x:~/vtest$ ls
test.v
linux1@ubu2410s390x:~/vtest$ v test.v
linux1@ubu2410s390x:~/vtest$ ls
test  test.v
linux1@ubu2410s390x:~/vtest$ ./test
0
[-339742632, 2417648]
[3, 4]
linux1@ubu2410s390x:~/vtest$ uname -a
Linux ubu2410s390x 6.11.0-21-generic #21-Ubuntu SMP Wed Feb 19 15:17:34 UTC 2025 s390x s390x s390x GNU/Linux
```

**Comment by barracuda156** [comment]
I need an advice. I tried to add some basic powerpc support, but encountered two issues:

1. I need an updated `vc`, and it is unclear how to generate it correctly. I tried to patch it manually, matching my patch to V source, but I think it is incorrect, and something was missed: https://github.com/vlang/v/issues/24225
It does not matter if I build via MacPorts or directly, `vc` is not regenerated automatically, but simply fetched from its repo. That does not work for testing a modified source.

2. Running the build, I got `v1` built, and then the build failed down the road, claiming that closures are undefined. This is strange: it can fail to compile, or compile but fail to work, but I did define closures for the arch, FWIW. Apparently something is just not found, but I am not familiar with the build system and do not even see where intermediate build files are placed.
Also, why `windows` is mentioned here? I am building on macOS.
```
----- imported modules: ------
['strings', 'math.bits', 'strconv', 'builtin', 'strings.textscanner', 'term.termios', 'time', 'os.cmdline', 'v.token', 'v.dotgraph', 'hash', 'sync.stdatomic', 'hash.fnv1a', 'encoding.utf8.validate', 'sync.threads', 'rand.buffer', 'flag', 'semver', 'os', 'v.errors', 'os.filelock', 'v.depgraph', 'rand.seed', 'term', 'v.help', 'v.util.version', 'v.vcache', 'v.util.recompilation', 'v.util.vflags', 'runtime', 'v.vmod', 'v.cflag', 'rand.config', 'rand.wyrand', 'v.pkgconfig', 'rand', 'v.pref', 'sync', 'v.util', 'sync.pool', 'v.ast', 'v.transformer', 'v.markused', 'v.type_resolver', 'v.scanner', 'v.ast.walker', 'v.checker', 'v.parser', 'v.callgraph', 'v.gen.c', 'v.builder', 'v.builder.cbuilder', 'main']
-------------------------------
0.000    ms Builder.resolve_deps
0.000    ms Builder.parse_imports
0.000    ms SCAN
0.000    ms PARSE
0.000    ms ALL_FRONT_STAGES
0.000    ms Checker.generic_insts_to_concrete
inferred `v.pref.Subsystem__static__from[string]`
inferred `work_on_items[&ast.File]`
inferred `work_on_items[string]`
0.000    ms CHECK
0.000    ms TRANSFORM
0.000    ms Table.complete_interface_check
0.000    ms all_global_decl
0.000    ms MARKUSED
0.000    ms Gen.write_sorted_fn_typesymbol_declaration
0.000    ms Gen.sort_structs
0.000    ms cgen init
gen fn `v.token.new_keywords_matcher_trie` for type `int`
gen fn `v.token.new_keywords_matcher_trie` for type `v.token.Kind`
gen fn `v.pref.Subsystem__static__from` for type `string`
gen fn `work_on_items` for type `v.ast.File`
gen fn `work_on_items` for type `string`
gen fn `get_item` for type `v.ast.File`
gen fn `get_item` for type `string`
gen fn `get_results` for type `os.Result`
gen fn `get_results_ref` for type `v.gen.c.Gen`
0.000    ms cgen serial processing
0.000    ms Gen.sort_globals_consts
0.000    ms cgen unification
0.000    ms Gen.write_init_function
0.000    ms Gen.interface_table
0.000    ms cgen common
0.000    ms C GEN
builder.cc() pref.out_name='/opt/local/var/macports/build/_opt_local_ppcports_lang_vlang/vlang/work/v-0.4.10/v2'
build_thirdparty_obj_files: v.ast.cflags: [CFlag{ name: "-I" value: "/opt/local/include/LegacySupport" mod: "time" os: "" cached: "" }, CFlag{ name: "-I" value: "/opt/local/var/macports/build/_opt_local_ppcports_lang_vlang/vlang/work/v-0.4.10/thirdparty/stdatomic/nix" mod: "sync.stdatomic" os: "" cached: "" }, CFlag{ name: "-l" value: "pthread" mod: "sync.threads" os: "" cached: "" }, CFlag{ name: "-l" value: "ws2_32" mod: "os" os: "windows" cached: "" }]
cc() isprod=true outname=/opt/local/var/macports/build/_opt_local_ppcports_lang_vlang/vlang/work/v-0.4.10/v2
> C compiler cmd: '/opt/local/bin/gcc-mp-14' '@/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c.rsp'
> C compiler response file "/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c.rsp":
-Os -arch ppc  -Os -Os -fwrapv -g -no-pie -O3 -flto -o '/opt/local/var/macports/build/_opt_local_ppcports_lang_vlang/vlang/work/v-0.4.10/v2' -I "/opt/local/include/LegacySupport" -I "/opt/local/var/macports/build/_opt_local_ppcports_lang_vlang/vlang/work/v-0.4.10/thirdparty/stdatomic/nix" "/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c" -std=c99 -D_DEFAULT_SOURCE -Wl,-export_dynamic -lpthread -Wl,-headerpad_max_install_names -latomic -L/opt/local/lib -lMacportsLegacySupport -arch ppc 
0.000    ms C gcc-mp-14
======== Output of the C Compiler (/opt/local/bin/gcc-mp-14) ========
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function '__closure_alloc':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11714:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_alloc'?
11714 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11830:40: note: in expansion of macro '_CLOSURE_SIZE'
11830 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11714:51: note: each undeclared identifier is reported only once for each function it appears in
11714 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11830:40: note: in expansion of macro '_CLOSURE_SIZE'
11830 |         int remaining = _V_page_size / _CLOSURE_SIZE;
      |                                        ^~~~~~~~~~~~~
In file included from /usr/include/string.h:148,
                 from /opt/local/include/LegacySupport/string.h:27,
                 from /tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:875:
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function '__closure_init':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11868:30: error: '__CLOSURE_GET_DATA_BYTES' undeclared (first use in this function); did you mean '__CLOSURE_GET_DATA'?
11868 |         memcpy(_closure_ptr, __CLOSURE_GET_DATA_BYTES, sizeof(__CLOSURE_GET_DATA_BYTES));
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11714:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_init'?
11714 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11871:25: note: in expansion of macro '_CLOSURE_SIZE'
11871 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function '__closure_create':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11714:51: error: '__closure_thunk' undeclared (first use in this function); did you mean '__closure_init'?
11714 | #define _CLOSURE_SIZE (((2*sizeof(void*) > sizeof(__closure_thunk) ? 2*sizeof(void*) : sizeof(__closure_thunk)) + sizeof(void*) - 1) & ~(sizeof(void*) - 1))
      |                                                   ^~~~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:11883:25: note: in expansion of macro '_CLOSURE_SIZE'
11883 |         _closure_ptr += _CLOSURE_SIZE;
      |                         ^~~~~~~~~~~~~
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'array_free':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:19232:40: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
19232 |         u8* mblock_ptr = ((u8*)((u64)(((u64)(a->data)) - ((u64)(a->offset)))));
      |                                        ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:19232:27: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
19232 |         u8* mblock_ptr = ((u8*)((u64)(((u64)(a->data)) - ((u64)(a->offset)))));
      |                           ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'ptr_str':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:20133:32: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
20133 |         string buf1 = u64_hex(((u64)(ptr)));
      |                                ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'char_str':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:20143:25: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
20143 |         return u64_hex(((u64)(cptr)));
      |                         ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'voidptr_str':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:20370:51: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
20370 |         return string__plus(_SLIT("0x"), u64_hex(((u64)(nn))));
      |                                                   ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'byteptr_str':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:20373:51: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
20373 |         return string__plus(_SLIT("0x"), u64_hex(((u64)(nn))));
      |                                                   ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'charptr_str':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:20376:51: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
20376 |         return string__plus(_SLIT("0x"), u64_hex(((u64)(nn))));
      |                                                   ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'StrIntpData_process_str_intp_data':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:23260:62: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
23260 |                                         string ss = u64_hex(((u64)(data->d.d_vp)));
      |                                                              ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'v__ast__ComptTimeConstValue_voidptr':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34928:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34928 |                 _option_ok(&(voidptr[]) { ((voidptr)(((i64)((*val._i8))))) }, (_option*)(&_t1), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34933:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34933 |                 _option_ok(&(voidptr[]) { ((voidptr)(((i64)((*val._i16))))) }, (_option*)(&_t2), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34938:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34938 |                 _option_ok(&(voidptr[]) { ((voidptr)(((i64)((*val._i32))))) }, (_option*)(&_t3), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34943:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34943 |                 _option_ok(&(voidptr[]) { ((voidptr)(((i64)((*val._i64))))) }, (_option*)(&_t4), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34948:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34948 |                 _option_ok(&(voidptr[]) { ((voidptr)(((u64)((*val._u8))))) }, (_option*)(&_t5), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34953:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34953 |                 _option_ok(&(voidptr[]) { ((voidptr)(((u64)((*val._u16))))) }, (_option*)(&_t6), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34958:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34958 |                 _option_ok(&(voidptr[]) { ((voidptr)(((u64)((*val._u32))))) }, (_option*)(&_t7), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34963:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34963 |                 _option_ok(&(voidptr[]) { ((voidptr)(((u64)((*val._u64))))) }, (_option*)(&_t8), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:34968:44: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
34968 |                 _option_ok(&(voidptr[]) { ((voidptr)(((u64)((*val._rune))))) }, (_option*)(&_t9), sizeof(voidptr));
      |                                            ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'v__ast__ComptTimeConstValue_i64':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:35056:40: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
35056 |                 _option_ok(&(i64[]) { ((i64)((*val._voidptr))) }, (_option*)(&_t13), sizeof(i64));
      |                                        ^
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c: In function 'v__ast__ComptTimeConstValue_u64':
/tmp/v_502/v2.ZZZZZZZZ0RRPHTVV836HC9AYAW.tmp.c:35176:40: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
35176 |                 _option_ok(&(u64[]) { ((u64)((*val._voidptr))) }, (_option*)(&_t12), sizeof(u64));
      |                                        ^
=====================================================================
/opt/local/bin/gcc-mp-14
=========

0.000    ms Checker.struct_init
0.000    ms Checker.struct_decl
0.000    ms Checker.struct setting default_expr_typ
make: *** [all] Bus error
```

**Comment by tankf33der** [comment]
@xet7 do you have access to ppc64le hardware ?

**Comment by barracuda156** [comment]
@tankf33der There is no `ppc64le` hardware, AFAIK. All powerpc is bi-endian. `ppc64le` is a software pick.

**Comment by xet7** [comment]
@tankf33der 

I did have remote access to ppc64le hardware, but then OpenPower MiniCloud was closed:

https://github.com/wekan/wekan/wiki/ppc

Currently I have:
- PPC hardware
  - PowerBook G4 Nvidia laptop (1 or 2 GB RAM), that can run MacOS Leopard, https://t2sde.org Linux, OpenBSD etc. Currently OpenBSD installed, because graphical desktop works with NVidia. With T2SDE, starting graphical desktop did not work, and I did not yet figure out how to get it booting, only way to boot was via reinstall.
  - PowerBook G4 Mini 1 GB RAM, that can run MacOS Leopard, MorphOS, Linux, etc. Currently running MorphOS
- RISC-V hardware
  - LicheePi4a: 8 GB RAM
  - MangoPi MPi-MQ1PH: 1 GB RAM
- Amiga 500 68000, Amiga 1200 68060 16+2 MB RAM
- Some armhf/armv7/arm64 hardware
- Remote ssh access to s390x Ubuntu/RHEL/SuSE
- Remote access to some clouds, I'll look can I find any sparc64

**Comment by xet7** [comment]
@tankf33der

For POWER8/POWER9/POWER10, maybe ask from https://openpowerfoundation.org/hub/ ?

Related https://news.ycombinator.com/item?id=44087707

**Comment by barracuda156** [comment]
@medvednikov Why did we close this? Neither ppc nor ppc64 are implemented yet.

**Comment by tankf33der** [comment]
@barracuda156 - The very semi-official position is this — the complexity of porting to a system without access to it + only ElfV2 ABI of the system will be supported. It will be necessary to maintain a balance between the number of users and the effort spent. It's like with `riscv32` — it's been in the project for a long time and works, but everyone is too lazy to even check and fix the subject of this issue.

**Comment by barracuda156** [comment]
@tankf33der There is no demands whatsoever to the upstream or other contributors. But it is desirable to have the issue opened, so that whomever is interested can see that someone else also cares.

I think this should be re-opened.

P. S. Remote access to the hardware is doable, if that is an obstacle.

**Comment by tankf33der** [comment]
@barracuda156 do you have ppc64 with ELFv2 ABI ?
```
gcc -dM -E - < /dev/null | grep -E "CALL|ELF"
```
There should be no `_CALL_ELF` variable.

**Comment by barracuda156** [comment]
@tankf33der My ppc and ppc64 machines are on Darwin. My only Linux installation is riscv64.

**Comment by xet7** [comment]
I also have hardware with many different CPU types. Please reopen this issue.

**Comment by JalonSolov** [comment]
It would be better to open separate issues for each one that still needs support.  That way, as support is added, those individual issues can be closed, leaving the uncompleted ones open.

**Comment by barracuda156** [comment]
@JalonSolov Fair enough, separate issues are good.

**Comment by tankf33der** [comment]
Yesterday I ported vlang to `ppc64` ELF v1 ABI and I also have a working version for `sparc64`.
Both versions can pass: `VTEST_ONLY=closure v test vlib`.
I will make PRs asap.

**Comment by tankf33der** [comment]
@barracuda156 @xet7 - `ppc64` and `sparc64` closure support in `master`.

**Comment by barracuda156** [comment]
@tankf33der Ok, so Darwin ppc/ppc64 left. I recall last time I got lost in the procedure how to generate sources to test, since simple patching was not enough. Could you refer me to a procedure or maybe help me to set it up? I can probably handle the implementation, now that Linux version is there, and got hardware to test, but I am not familiar with vlang build setup. (I use MacPorts’s port, which works; what I need is a way to get my patches into the code that gets compiled.)

**Comment by JalonSolov** [comment]
ppc64 is there now.  That's one of the 2 he mentioned in his last comment.

**Comment by barracuda156** [comment]
@JalonSolov I have seen that, but Darwin ABI differs from ELF. (It may not matter for some trivial code, but usually it does.)

**Comment by JalonSolov** [comment]
Of course it differs.  <sigh>

Using the MacPorts package is a good way to stay behind... there are exactly **ZERO** official V packages (unless you count the ZIP files you can download from github), and most of the packages that exist for any platform were created and never maintained.

We always recommend installing V from source, and that is doubly true if you want to work on V itself.  https://github.com/vlang/v?tab=readme-ov-file#installing-v-from-source

Just grep for `ppc64` in the source to find where you may need to make changes.

**Comment by barracuda156** [comment]
Sure enough I install from source, at the very minimum that is literally the only option in case of PowerPC. MacPorts is not Homebrew, it handles source builds.
